### PR TITLE
DAOS-12694 cart: Incorporate OPX enablement patch

### PR DESCRIPTION
**Pull request by frostedcmos**
- Incorporate OPX patch from Cornelis Networks.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by soumagne — APPROVED** [review]

**Review by jgmoore-or — APPROVED** [review]

File: src/cart/crt_hg.c [hunk 1]
```diff
@@ -677,10 +730,10 @@ crt_get_info_string(bool primary, int provider, char **string, int ctx_idx)
 	}
 
 out:
-	if (*string == NULL)
+	if (rc == DER_SUCCESS && *string == NULL)
```
Comment (charlesshereda) [inline src/cart/crt_hg.c:733]: I assume that we're the ones who did this, but it seems a bit odd to me that we're looking for a success return code and null pointer to then return a NOMEM return code. I'll separately look at crt_get_opx_info_string().

File: src/cart/crt_hg.c [hunk 2]
```diff
@@ -623,13 +628,54 @@ d_list_t
...
+			   provider, domain, hfi,
+			   start_port + ctx_idx);
+	else
+		D_ASPRINTF(*string, "%s://%s/%s:%d:%d",
+			   provider, domain, ip, hfi,
+			   start_port + ctx_idx);
+
+out:
+	if (!rc && *string == NULL)
+		return -DER_NOMEM;
```
Comment (charlesshereda) [inline src/cart/crt_hg.c:666]: Seems from this as if we should already be returning -DER_NOMEM for all out-of-memory cases. Not sure how we could get into the case that I commented on below but perhaps other providers are?

**Comment by bwilsoncn** [comment]
Mercury ctest and CaRT tests are completed, confirmed .is_contig=false will be needed, at least initially.  Daos_test suite in progress.

**Comment by charlesshereda** [comment]
@frostedcmos will you please add @bwilsoncn as reviewer to this PR as well?

**Comment by charlesshereda** [comment]
@frostedcmos Brian is waiting to approve until he's tested the Daos_test suite - does that make sense to you or should we proceed before that suite has been fully tested?

**Comment by frostedcmos** [comment]
> @frostedcmos Brian is waiting to approve until he's tested the Daos_test suite - does that make sense to you or should we proceed before that suite has been fully tested?

I would prefer we get this landed soon, as to avoid having to remerge this PR. If any subsequent changes need based on daos_test findings then we can handle it in a separate PR.

File: src/cart/crt_hg.c (hunk 1 above)
Comment (MosheAri) [reply src/cart/crt_hg.c:733]: Peace Charles S,
I was being overly cautious just ensuring that the result was in alignment with the string allocation.

**Comment by bwilsoncn** [comment]
I've completed retesting equivalent to before the v2.4 rebase, looks good for merge.

**Review by charlesshereda — APPROVED** [review]

**Comment by bwilsoncn** [comment]
@frostedcmos just to make sure it doesn't fall through the cracks, we discussed a few approaches to handling the "opx" provider string but I don't remember reaching a conclusion. I'm still patching crt_hg.c with .nad_alt_str: "opx"

**Comment by kjacque** [comment]
> thanks, i believe @kjacque has already made changes to the control plane in master to allow ofi+opx passing as a string, so no need to have .nad_alt_str for it anymore. I'll refresh the patch today and will either add or remove .nad_alt_str = 'opx' depending on whether Kris confirms that the issue is addressed.

The issue is indeed resolved in the DAOS master branch.

**Pull request merged**